crates/alloy-provider/src/lib.rs

```diff
@@ -0,0 +1,1460 @@
+#[derive(Debug, Clone, Default)]
+pub struct AlloyRethProviderConfig {
+    /// Whether to compute state root when creating execution outcomes
+    pub compute_state_root: bool,
+}
+
+impl AlloyRethProviderConfig {
+    /// Sets whether to compute state root when creating execution outcomes
+    pub const fn with_compute_state_root(mut self, compute: bool) -> Self {
+        self.compute_state_root = compute;
+        self
+    }
+}
+
+/// A provider implementation that uses Alloy RPC to fetch state data
+///
+/// This provider implements reth's `StateProviderFactory` and related traits,
+/// allowing it to be used as a drop-in replacement for database-backed providers
+/// in scenarios where RPC access is preferred (e.g., testing).
+///
+/// The provider type is generic over the network type N (defaulting to `AnyNetwork`),
+/// but the current implementation is specialized for `alloy_network::AnyNetwork`
+/// as it needs to access block header fields directly.
+#[derive(Clone)]
+pub struct AlloyRethProvider<P, Node, N = alloy_network::AnyNetwork>
+where
+    Node: NodeTypes,
+{
+    /// The underlying Alloy provider
+    provider: P,
+    /// Node types marker
+    node_types: std::marker::PhantomData<Node>,
+    /// Network marker
+    network: std::marker::PhantomData<N>,
+    /// Broadcast channel for canon state notifications
+    canon_state_notification: broadcast::Sender<CanonStateNotification<PrimitivesTy<Node>>>,
+    /// Configuration for the provider
+    config: AlloyRethProviderConfig,
+    /// Cached chain spec
+    chain_spec: Arc<Node::ChainSpec>,
+}
+
+impl<P, Node: NodeTypes, N> std::fmt::Debug for AlloyRethProvider<P, Node, N> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("AlloyRethProvider").field("config", &self.config).finish()
+    }
+}
+
+impl<P, Node: NodeTypes, N> AlloyRethProvider<P, Node, N> {
+    /// Creates a new `AlloyRethProvider` with default configuration
+    pub fn new(provider: P) -> Self
+    where
+        Node::ChainSpec: Default,
+    {
+        Self::new_with_config(provider, AlloyRethProviderConfig::default())
+    }
+
+    /// Creates a new `AlloyRethProvider` with custom configuration
+    pub fn new_with_config(provider: P, config: AlloyRethProviderConfig) -> Self
+    where
+        Node::ChainSpec: Default,
+    {
+        let (canon_state_notification, _) = broadcast::channel(1);
+        Self {
+            provider,
+            node_types: std::marker::PhantomData,
+            network: std::marker::PhantomData,
+            canon_state_notification,
+            config,
+            chain_spec: Arc::new(Node::ChainSpec::default()),
+        }
+    }
+
+    /// Helper function to execute async operations in a blocking context
+    fn block_on_async<F, T>(&self, fut: F) -> T
+    where
+        F: Future<Output = T>,
+    {
+        tokio::task::block_in_place(move || Handle::current().block_on(fut))
```

Comment:
i see, this `block_in_place` should be ok for testing/development, if at any point we need more performance or responsiveness under heavy load we could consider other async approaches